### PR TITLE
Coupons: Simplify title for Edit Coupon view and enable updating discount type

### LINKED ISSUE #6912: Coupons: Update Edit coupon flow with shorter title and option to change discount type
The current title for this screen makes it hard to display the content fully on small devices.
We should change the title to "Edit Coupon" and add a new field to edit discount type:

Screen Shot 2022-05-20 at 12 21 18

Internal discussion: p1653017206944579-slack-C0354HSNUJH

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #6912 

### Description

Currently, we're specifying discount type on the title of the Edit Coupon view, which is lengthy that can be cut off on small devices. Also, this indicates that the coupon's discount type cannot be updated, which doesn't align with the feature on WP Admin.

This PR addresses both of the issues:
- Make the title of the Edit Coupon view generic by using "Edit Coupon" or "Create Coupon" - depending on the flow the screen is displaying.
- Enable updating discount type:
  - Since we're using the custom bottom sheet design to present the list of discount types with descriptions, a new `BottomSheetListSelectorCommand` type is added for discount type.
  - Update `ImageAndTitleAndTextTableViewCell` with a checkmark for the accessory view if the item it represents is selected.
  - Present the discount type bottom sheet from the Edit coupon view. Since we want to display the bottom sheet that covers just half the screen on iPhone while showing a popover pointing from the correct view, my solution was to present the sheet the UIKit way on iPhone and embed it in a popover on iPad. This should be updated with a [`UISheetPresentationController`](https://developer.apple.com/documentation/uikit/uisheetpresentationcontroller) wrapper when we drop support for iOS 14.

### Testing instructions

- Make sure to enable coupon management in experimental features.
- Navigate to Menu > Coupons > select a coupon with a supported discount type.
- Select the ellipsis button > Edit Coupon.
- Notice that the presented screen's title is "Edit Coupon", and a new field for updating discount type is on the top.
- Tap Discount type, and notice that a bottom sheet is presented if you're on an iPhone. On iPads, the content is displayed in a popover presented from the Discount type row.
- When switching between discount types, the Amount field should be updated accordingly. Saving the updates should succeed.

### Screenshots

On iPhone:

https://user-images.githubusercontent.com/5533851/170169510-e100b806-41f4-438a-9d64-de4d9eb79571.mp4

On iPad:

https://user-images.githubusercontent.com/5533851/170171297-cf46f645-78b9-4baf-b97c-61eca05d4786.mov

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/BottomSheet/BottomSheetListSelector.swift [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        self.command = command
+        self.onDismiss = onDismiss
+    }
+
+    func makeUIViewController(context: Context) -> BottomSheetListSelectorViewController<Command, Model, Cell> {
+        return BottomSheetListSelectorViewController(viewProperties: viewProperties, command: command, onDismiss: onDismiss)
+    }
+
+    func updateUIViewController(_ uiViewController: BottomSheetListSelectorViewController<Command, Model, Cell>, context: Context) {
+        // no=op
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/BottomSheet/BottomSheetListSelector.swift]: super-nit: 
```suggestion
        // no-op
```

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift [hunk 2]
```diff
@@ -99,6 +118,25 @@ struct AddEditCoupon: View {
...
 
                             Group {
+                                TitleAndValueRow(title: Localization.discountType,
+                                                 value: viewModel.discountTypeValue,
+                                                 selectionStyle: .disclosure, action: {
+                                    if idiom == .pad {
+                                        showingDiscountType.toggle()
+                                    } else {
+                                        discountTypeHandler(viewProperties)
+                                    }
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:129]: nit: I haven't worked with SwiftUI for a while and tried to remember the iOS 14 issue here 😅 could you briefly explain the issue? I don't see an issue with bottom sheet in the SwiftUI workarounds doc P91TBi-7F9-p2, could we add it here for future reference? When we drop iOS 14 support, we can update all the workarounds from iOS 14 in the doc.

Pretty minor but just wanted to mention: on iPad, the bottom sheet looks a bit tall in split view (iOS 15.5 simulator):

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/DiscountTypeBottomSheetListSelectorCommand.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import Foundation
+import Yosemite
+
+/// `BottomSheetListSelectorCommand` for selecting a discount type for a coupon.
+///
+final class DiscountTypeBottomSheetListSelectorCommand: BottomSheetListSelectorCommand {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/DiscountTypeBottomSheetListSelectorCommand.swift:6]: nit: some unit tests might be helpful, like for the initial `selected` and selection behavior 🙂  (there are probably some tests for other `BottomSheetListSelectorCommand` subclasses)

**Review by jaclync — APPROVED** [review]
LGTM! :shipit: 

Non-blocking question:
Not sure if this was mentioned before, I guess I'm so used to the "Done"/"Save" CTA in the navigation bar to save an item in the app that I felt a bit lost when I didn't see one in the edit coupon screen. Alternatively, the CTA is at the end of the edit coupon screen which isn't visible without scrolling. Just wondering if we've discussed why we have a different UX to save a coupon?

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:129]: The issue with the bottom sheet was discussed here: p1653302908311579-slack-C02TD63FZA8.

So we want to be able to present a bottom sheet that covers only half of the screen, and this is not supported natively on iOS 14 yet, so I had to do this workaround. I'll include it on the P2 page when I finish this PR - thanks for reminding me!

I can't seem to repro the split view case on iPad, could you please include more information on how you got this error, @jaclync?

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/DiscountTypeBottomSheetListSelectorCommand.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/DiscountTypeBottomSheetListSelectorCommand.swift:6]: Makes sense! I added tests in 68f4bd030ce0474c794a7e2d1558c194622d1539 and 21cd505e3e3e2079884cf8f0a134199f40e17e4f.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:129]: Thanks for the link to the discussion! It's a shame many SwiftUI APIs are only added in recent iOS versions for developers to come up with workarounds (the drawback of adopting a new framework before it's fully featured 😞 ). I don't think this is an error and it's most likely a minor use case, it's probably the popover falling back to action sheet when the horizontal size class changes to compact. Totally fine to ship it with this, I think we had similar issues where the share sheet is full height using SwiftUI without a good solution. Just adding a note to migrate to use the iOS 15+ `adaptiveSheetPresentationController` bottom sheet should be good 👍 

I just tested in Xcode 13.3.1 (was using Xcode 13.4 earlier) iPad Air gen 4, and here's a recording:

https://user-images.githubusercontent.com/1945542/170404944-525b38a5-0e68-4816-8931-30c5f49d362d.mov

**Review by jaclync — APPROVED** [review]
Thanks for adding the tests! Still looks good 👍

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:129]: I've added a comment in 072e8a8455bdfb18a124a0b9b70ece415a0a1a76 to remove the workaround when we drop iOS 14, and also updated the SwiftUI workaround doc with a new section for this.

**Pull request merged**